stage0/src/apic.rs

```diff
@@ -0,0 +1,249 @@
+//
+// Copyright 2023 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use core::arch::x86_64::__cpuid;
+
+use bitflags::bitflags;
+use x86_64::{registers::model_specific::Msr, PhysAddr};
+
+bitflags! {
+    #[derive(Clone, Copy, Debug)]
+    pub struct ApicBaseFlags: u64 {
+        /// APIC Enable
+        const AE = (1 << 11);
+
+        /// x2APIC Mode Enable
+        const EXTD = (1 << 10);
+
+        /// Boot Strap CPU Core
+        const BSC = (1 << 8);
+    }
+}
+
+/// The APIC Base Address Register.
+pub struct ApicBase;
+
+impl ApicBase {
+    pub const MSR: Msr = Msr::new(0x0000_001B);
+
+    fn read_raw() -> u64 {
+        unsafe { Self::MSR.read() }
+    }
+
+    fn write_raw(value: u64) {
+        let mut msr = Self::MSR;
+        unsafe { msr.write(value) }
```

Comment:
Curious about these. The docs say that the caller must ensure that there's no unsafe side effects when calling these methods on the MSR. Which unsafe side effects could these be? How do we know they're not occuring here?

Reply:
Yes, I should have been more thorough with my safety comments :)

In short, we only support x2APIC which makes those MSRs available, and we do check that the CPU supports x2APIC. If the CPU doesn't support x2APIC, `Lapic::enable()` will give you an error. Thus, accessing those MSRs is safe.